core/src/main/java/io/cloudevents/builder/CloudEventBuilder.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2018-Present The CloudEvents Authors
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.cloudevents.builder;
```

Comment:
In lue of Java Module System that does not allow the same package to be defined in multiple modules, I'd recommend to put everything in `core` to `io.cloudevents.impl` sub-packages. So that, if we ever decide to expose builders in the API, we will no have a problem of `io.cloudevents.builder` and similar being defined in `core` and not `api`

Reply:
ok I'm sorry but you find me a little unprepared on this. Why does the java module system should break? There are no packages that are defined both in core and api at the moment.

And any reason why `impl` more than `core` (so it reflects the module name)?

Reply:
api defines:

- `io.cloudevents`
- `io.cloudevents.lang`
- `io.cloudevents.types`
- `io.cloudevents.visitor`

while core defines:

- `io.cloudevents.impl`
- `io.cloudevents.builder`
- `io.cloudevents.message`
- `io.cloudevents.format`
- `io.cloudevents.extensions`
- `io.cloudevents.v1`
- `io.cloudevents.v03`

Reply:
If for some reason someday we want to move `message` or `builder` to api we move the whole package, i carefully avoided to put both in `message` and `builder` only the interfaces and not the implementations

Reply:
Does that makes sense?

Reply:
Yes, sorry, I wasn't clear enough. Currently, there is no issue indeed.

But, let's say one day we decide to add `extensions` sub-package to `api`.
We won't be able to use `io.cloudevents.extensions` because it is already defined in `core`, so we would need to find another (probably hacky) name for it.

If we put everything in `core` into `io.cloudevents.impl.*`, we protected future ourselves from having to fight with JPMS.
As a neat side effect, It also makes it clear that any import from `io.cloudevents.impl.*` is the implementation detail.

Reply:
> we move the whole package

this may break projects that have newer `api.jar` but older `core.jar` (can easily happen when a dependency defines only `api` as its dependency, and uses a newer one)

Reply:
Ok but if we ever move a package to api, it's an _"api"_ package, so it won't contain any implementation (in theory) and you want to move the whole package (excluding subpackages), not just a couple of interfaces splitting the package between api and core. For the particular case of `message` and `builder`, I don't think splitting it moving a part to `api` and keeping a part in `core` makes sense. Do you foresee one particular module that could require such "splitting"?

The case of `extensions` it's interesting because extensions IMO should contain only the "extension pojos", but now it contains that `ExtensionsParser` that probably needs to be moved somewhere else. For that particular case I've opened the issue #122 to discuss about Extensions. I think the same issue applies to `format` package.

> As a neat side effect, It also makes it clear that any import from io.cloudevents.impl.* is the implementation detail.

Yeah but it has some interfaces too that other modules uses, that's why I prefer `core`.

Reply:
> this may break projects that have newer api.jar but older core.jar (can easily happen when a dependency defines only api as its dependency, and uses a newer one)

Ok but this is the kind of change done between majors, but yes i see the point

Reply:
let me try to move to `io.cloudevents.core`

Reply:
> Yeah but it has some interfaces too that other modules uses

Ideally, the modules should also define their own unique packages.

Just sharing some **very** painful experience working on a library that did not expect JPMS and such limitations to happen 😂

Reply:
Check it now